### ISSUE TITLE
[🚀 Feature]: Download grid within Bindings

### ISSUE THREAD

**Issue by titusfortner** [SeleniumHQ/selenium#12305]
### Status
* Ruby: : #12299
* Python: #15666
* Java: (n/a - this can/should be done directly with the remote jar)
* JS: 
* .NET: 

### Feature and motivation

Ruby has always had a wrapper that will download a specified version of the grid so that it can be started for running tests against the server. This isn't necessary now that we are using bazel for the tests. But. It would be nice to have this for our [example code](https://github.com/SeleniumHQ/seleniumhq.github.io/tree/trunk/examples). 
Now that it is [supported by Selenium Manager](https://github.com/SeleniumHQ/selenium/pull/12281), it should be pretty straightforward to support.

### Usage example

This is an example of the Ruby implementation

```
      server = Selenium::Server.get(:latest, background: true)
      server.start

      driver = Selenium::WebDriver.for :remote, url: server.webdriver_url, options: Selenium::WebDriver::Options.chrome
      
      driver.quit
      server.stop
```

**Comment by titusfortner** [comment]
Want to wait on this until we can get #13022 sorted

**Comment by Delta456** [comment]
Are we still going with this?

**Comment by diemol** [comment]
This is something Titus wants to implement. The last time we met, we decided it was a very low priority.

**Comment by cgoldberg** [comment]
I think this is still worth doing.

I didn't even realize Selenium Manager could do this. Right now I'm using this dubious shell function in my `~/.bashrc` to grab it from GitHub release assets:

```
# run selenium server in standalone mode (port 4444)
# download server first with github cli if it's not found
selenium-server () {
    local sel_home="${HOME}/code/selenium"
    local jar="selenium-server.jar"
    if [ ! -x "$(command -v gh)" ]; then
        err "github cli is not installed"
        return 1
    fi
    if [ ! -d "${sel_home}" ]; then
        err "can't find selenium repo at: ${sel_home}"
        return 1
    fi
    cd ${sel_home}
    if [ ! -f "${jar}" ]; then
        gh release download --pattern=selenium-server*.jar --output=${jar}
    fi
    cp ${jar} "${sel_home}/java/src/org/openqa/selenium/grid/selenium_server_deploy.jar"
    java -jar ${jar} standalone
}
```

**Comment by titusfortner** [comment]
This is the code I'm currently using in Ruby:
```
    jar = Selenium::WebDriver::SeleniumManager.binary_paths('--grid')['driver_path']
```

Should work similarly in Python. The idea for this issue is to create a basic wrapper for it.
Ruby has had a class that does this the old way that I planned to update, but there are implementation issues with how it is currently namespaced that kept me from just doing it.

**Comment by cgoldberg** [comment]
I'll look into adding something for Python. I just need to figure out where it should be located.

Downloading the server should be easy... but running it will be more complicated. (make sure java is installed, launch process in the background, terminate the process, etc)

**Comment by nvborisenko** [comment]
I think the grid is more related to orchestration, and in my opinion it is mistake trying to delegate orchestration to bindings (bindings are just libraries).

OK, binding will download standalone jar... Hmm, jar?!.. Hmm, standalone?!.. OK, if java is required, then the following feature request would be "support docker". I think we are trying to jump into area where **bindings are not responsible for**.

I propose to understand the motivation, review use cases, and probably we find good solution of this feature request. For me it is pretty simple to:

```cmd
selenium-manager grid start --port 5555
```

That's it! Bindings are not here at all. If we enhance how we are shipping SM, then it could be even simplified for end user... Full flow could be:

```cmd
dotnet install -g Selenium.Manager # variant of npx install
selenium-manager grid start --port 5555 # downloads and starts
```

**Comment by titusfortner** [comment]
> running it will be more complicated

Oh, right, yes, this is what Ruby code has had forever: https://github.com/SeleniumHQ/selenium/blob/61fa8723a219e28f6a08808da8aff1319a579efa/rb/lib/selenium/server.rb

**Comment by titusfortner** [comment]
@nvborisenko for example, how do you propose we run this properly?

https://github.com/SeleniumHQ/seleniumhq.github.io/blob/trunk/examples/dotnet/SeleniumDocs/Drivers/RemoteWebDriverTest.cs

**Comment by titusfortner** [comment]
To orchestrate it, we'd need to add a start / stop to the command line run of the GitHub actions, and hard code the ports. And we only need it for that one test, it doesn't need to be running the entire time of the suite execution.

**Comment by cgoldberg** [comment]
> in my opinion it is mistake trying to delegate orchestration to bindings (bindings are just libraries).

You might be right. I'll still implement it in Python, but it might be better as a 3rd party package (if this doesn't already exist). I'll have to give it some thought.

**Comment by titusfortner** [comment]
I guess I'm a little biased because there has been an easy way to do this in both Java and Ruby pretty much from the beginning.  I'm the one that pushed for this with the idea that we'd be able to do this with a small class in Python & .NET to match what is possible with Ruby & Java without much overhead.

**Comment by titusfortner** [comment]
> ### Status
> * Ruby: : [[rb] implement downloading grid with Selenium Manager #12299](https://github.com/SeleniumHQ/selenium/pull/12299)
> * Python:
> * Java: (not necessary, can use the jar directly)
> * JS:
> * .NET:
> 
> ### Feature and motivation
> Ruby has always had a wrapper that will download a specified version of the grid so that it can be started for running tests against the server. This isn't necessary now that we are using bazel for the tests. But. It would be nice to have this for our [example code](https://github.com/SeleniumHQ/seleniumhq.github.io/tree/trunk/examples). Now that it is [supported by Selenium Manager](https://github.com/SeleniumHQ/selenium/pull/12281), it should be pretty straightforward to support.
> 
> ### Usage example
> This is an example of the Ruby implementation
> 
> ```
>       server = Selenium::Server.get(:latest, background: true)
>       server.start
> 
>       driver = Selenium::WebDriver.for :remote, url: server.webdriver_url, options: Selenium::WebDriver::Options.chrome
>       
>       driver.quit
>       server.stop
> ```

**Comment by nvborisenko** [comment]
> To orchestrate it, we'd need to add a start / stop to the command line run of the GitHub actions, and hard code the ports. And we only need it for that one test, it doesn't need to be running the entire time of the suite execution.

What I see: it is your problem, not end users. Nobody came to us with issue like "how to download grid and start it programmatically". If in .NET user will write `SeleniumManager.StartGrid()` and this command fails with error like `dear user, please install java`. And after retry user sees error `dear user, you must add java to environment variables`. And even after that user sees `dear user, your java is not supported within grid`, this is bad experience.

User is responsible for infrastructure setup.

**Comment by cgoldberg** [comment]
@nvborisenko 
I guess based on that logic, we should remove Selenium Manager and just fail with "dear user, install the webdriver" like we used to. User is responsible for infrastructure setup.

**Comment by titusfortner** [comment]
Yes, and I solve my Selenium problems by updating Selenium codebase. :) 

I don't disagree with the philosophy really, but it would be super convenient for me to do demos and run tests outside of Selenium if we essentially made this class available to users and integrated it with Selenium Manager execution
https://github.com/SeleniumHQ/selenium/blob/0954ccac4c86e4ba81db65d255f1d04d8edd7b29/dotnet/test/common/Environment/RemoteSeleniumServer.cs

**Comment by titusfortner** [comment]
and if the choice is between doing a bunch of orchestration in multiple places for the things I'm doing, or creating my own package, or adding a single class to Selenium that we already have the code for, and then making sure the Selenium Manager class can handle it... Seems like an obvious choice for minimizing my total maintenance burden.

**Comment by nvborisenko** [comment]
> [@nvborisenko](https://github.com/nvborisenko) I guess based on that logic, we should remove Selenium Manager and just fail with "dear user, install the webdriver" like we used to. User is responsible for infrastructure setup.

I liked this comment, absolutely agree. But the difference is that `Selenium Manager` is native tool without any infrastructure dependencies, `Standalone Grid` requires a lot of pre-requisites. This is a key point why I don't see grid to be a dependency of any binding.

**Comment by titusfortner** [comment]
The only additional pre-requisite for it is JRE, which isn't really limiting, especially as it is directly related to the functionality being executed. I do get your general point, though.

**Comment by diemol** [comment]
I have never seen the use case for our users here. It makes sense for browser drivers to have Selenium Manager. It helps us as maintainers to run tests in CI, I don't see any other use case to be honest.

**Comment by cgoldberg** [comment]
> It helps us as maintainers to run tests in CI

It would help users run tests in their own CI (or anywhere else). Without this, you need some sort of shell script/command that differs depending on the platform you run on. It's not a "must have" feature, but it's a nice convenience that cuts down on external setup.

**Comment by diemol** [comment]
Most CI users rely on Docker-Selenium, or a combination of browser drivers directly + XVFB.

**Comment by diemol** [comment]
I am not saying not to do it; I only share my opinion. This is not a use case for the end user.

**Comment by nvborisenko** [comment]
It would help users to write in any language "test my application". Please let's us not forget about "Selenium automates browsers. That's it!".

**Comment by titusfortner** [comment]
You're right that it isn't how most users will want to do it, but
Docker is way overkill for what I want to be able to do. On that note, though, what is the intended use case for standalone mode in general? 

This is essentially, take our test code for starting the server and add version flexibility from selenium manager, and make it publicly accessible.

**Comment by nvborisenko** [comment]
I am still strongly disagree that binding should do this job. This is orchestration. Selenium is a library/binding.

**Comment by aguspe** [comment]
@titusfortner is this something we still want to do?
